### PR TITLE
KafkaIndexTask: Allow run thread to stop gracefully instead of interrupting

### LINKED ISSUE #3531: Kafka indexing fails to stop correctly in 0.9.2rc1
**Steps to reproduce:**

1. Create a [kafka indexer](http://druid.io/docs/0.9.1.1/development/extensions-core/kafka-ingestion.html)
2. Wait for a kafka index task to start
3. Emulate a `/stop` request sent from supervisor to worker, e.g. via
```
http post localhost:8100/druid/worker/v1/chat/index_kafka_Example_0f2f4bbb5f488f3_mjmpbfek/stop
```

**Expected behavior:**
Worker successfully stops

**Actual behavior:**
Worker process remains lingering until middle manager forcefully kills the process after timeout. The following exception is seen in the logs:

```
2016-10-04T13:16:56,181 ERROR [task-runner-0-priority-0] io.druid.indexing.overlord.ThreadPoolTaskRunner - Exception while running task[KafkaIndexTask{id=index_kafka_Example_0f2f4bbb5f488f3_mjmpbfek, type=index_kafka, dataSource=Example}]
java.util.concurrent.RejectedExecutionException: Got Interrupted while adding to the Queue
	at io.druid.concurrent.Execs$1.rejectedExecution(Execs.java:123) ~[druid-common-0.9.2-rc1.jar:0.9.2-rc1]
	at java.util.concurrent.ThreadPoolExecutor.reject(ThreadPoolExecutor.java:823) ~[?:1.8.0_102]
	at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1369) ~[?:1.8.0_102]
	at com.google.common.util.concurrent.MoreExecutors$ListeningDecorator.execute(MoreExecutors.java:484) ~[guava-16.0.1.jar:?]
	at java.util.concurrent.AbstractExecutorService.submit(AbstractExecutorService.java:134) ~[?:1.8.0_102]
	at com.google.common.util.concurrent.AbstractListeningExecutorService.submit(AbstractListeningExecutorService.java:58) ~[guava-16.0.1.jar:?]
	at io.druid.segment.realtime.appenderator.AppenderatorImpl.persistAll(AppenderatorImpl.java:377) ~[druid-server-0.9.2-rc1.jar:0.9.2-rc1]
	at io.druid.segment.realtime.appenderator.AppenderatorImpl.add(AppenderatorImpl.java:218) ~[druid-server-0.9.2-rc1.jar:0.9.2-rc1]
	at io.druid.segment.realtime.appenderator.FiniteAppenderatorDriver.add(FiniteAppenderatorDriver.java:199) ~[druid-server-0.9.2-rc1.jar:0.9.2-rc1]
	at io.druid.indexing.kafka.KafkaIndexTask.run(KafkaIndexTask.java:416) ~[?:?]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:436) [druid-indexing-service-0.9.2-rc1.jar:0.9.2-rc1]
	at io.druid.indexing.overlord.ThreadPoolTaskRunner$ThreadPoolTaskRunnerCallable.call(ThreadPoolTaskRunner.java:408) [druid-indexing-service-0.9.2-rc1.jar:0.9.2-rc1]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_102]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [?:1.8.0_102]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [?:1.8.0_102]
	at java.lang.Thread.run(Thread.java:745) [?:1.8.0_102]
2016-10-04T13:16:56,284 INFO [task-runner-0-priority-0] io.druid.indexing.overlord.TaskRunnerUtils - Task [index_kafka_Example_0f2f4bbb5f488f3_mjmpbfek] status changed to [FAILED].
2016-10-04T13:16:56,293 INFO [task-runner-0-priority-0] io.druid.indexing.worker.executor.ExecutorLifecycle - Task completed with status: {
  "id" : "index_kafka_Example_0f2f4bbb5f488f3_mjmpbfek",
  "status" : "FAILED",
  "duration" : 1811725
}
```

---

This does not happen every time, but in about 70% of the cases so far. This issue is quite similar to https://github.com/druid-io/druid/issues/3346, except that this time the lookups do not have anything to do with it.

As I understand, `KafkaIndexTask` attempts to persist everything to `FiniteAppenderatorDriver`, but since the peon is being shut down, submitting a task in `AppenderatorImpl.persistAll` gets rejected.

I do not yet see what change could have caused it, but we do not observe the issue in 0.9.1.1. Any ideas? This would be a show-stopper for upgrading to 0.9.2 for anyone using kafka ingestion.

### PR DESCRIPTION
**Pull request by dclim**
Doesn't fix #3531

### CODE REVIEW

**Comment by dclim** [comment]
Actually hm, I think we should still interrupt the thread if it's in the publishing state because if another replica completes before this one there's no point sitting around waiting for handoff.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java [hunk 1]
```diff
@@ -552,11 +559,45 @@ public boolean canRestore()
...
+        finally {
+          pauseLock.unlock();
+        }
+      } else {
+        log.warn("While stopping: failed to acquire pauseLock before timeout, interrupting run thread");
+        runThread.interrupt();
+        return;
+      }
+
+      if (pollRetryLock.tryLock(15, TimeUnit.SECONDS)) {
```
Comment (gvsmirnov) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java]: I'm not sure what the conventions are here, but would it make sense to extract the 15 seconds from here and the `pauseLock.tryLock` above, and perhaps de-hardcode?

Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java]: @gvsmirnov sounds reasonable to me, thanks for the review

**Comment by dclim** [comment]
@gianm you're right, there's actually two "issues" in #3531: the main problem is that the task doesn't end because a thread doesn't exit which is addressed in #3539, but the OP first posted a stack trace with `java.util.concurrent.RejectedExecutionException: Got Interrupted while adding to the Queue` which turned out to be unrelated to the process being hung. I submitted this PR earlier to make the task exit cleaner on a /stop, and then when OP mentioned the Kafka lookup extension I was able to repro it and submitted the fix to the real issue in #3539.

File: extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java [hunk 2]
```diff
@@ -161,6 +163,7 @@
   private final Condition shouldResume = pauseLock.newCondition();
   private final Lock pollRetryLock = new ReentrantLock();
   private final Condition isAwaitingRetry = pollRetryLock.newCondition();
+  private final Object statusLock = new Object();
```
Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java]: @gianm there's a big comment just above these lines that describes what most of the locks and conditions do, I'll add more details for pollRetryLock and statusLock

Comment (dclim) [inline extensions-core/kafka-indexing-service/src/main/java/io/druid/indexing/kafka/KafkaIndexTask.java]: @gianm added comments

**Comment by fjy** [comment]
👍

**Pull request merged**